packages/annotations/src/format/annotation.js

```diff
@@ -52,31 +57,104 @@ export function removeAnnotations( record ) {
+					start: i,
+				};
+			}
+
+			// Annotations refer to positions between characters.
+			// Formats refer to the character themselves.
+			// So we need to adjust for that here.
+			positions[ id ].end = i + 1;
+		} );
+	} );
+
+	return positions;
+}
+
+/**
+ * Updates annotations in the state based on positions retrieved from RichText.
+ *
+ * @param {Array}    annotations           The annotations that are currently applied.
+ * @param {Array}    positions             The current positions of the given annotations.
+ * @param {Function} removeAnnotation      Function to remove an annotation from the state.
+ * @param {Function} updateAnnotationRange Function to update an annotation range in the state.
+ */
+function updateAnnotationsWithPositions( annotations, positions, { removeAnnotation, updateAnnotationRange } ) {
+	annotations.forEach( ( currentAnnotation ) => {
+		const position = positions[ currentAnnotation.id ];
+		// If we cannot find an annotation, delete it.
+		if ( ! position ) {
+			// Apparently the annotation has been removed, so remove it from the state:
+			// Remove...
+			removeAnnotation( currentAnnotation.id );
+			return;
+		}
+
+		const { start, end } = currentAnnotation;
+		if ( start !== position.start || end !== position.end ) {
+			updateAnnotationRange( currentAnnotation.id, position.start, position.end );
+		}
+	} );
+}
+
 export const annotation = {
-	name,
+	name: FORMAT_NAME,
 	title: __( 'Annotation' ),
 	tagName: 'mark',
 	className: 'annotation-text',
 	attributes: {
 		className: 'class',
+		id: 'id',
 	},
 	edit() {
 		return null;
 	},
 	__experimentalGetPropsForEditableTreePreparation( select, { richTextIdentifier, blockClientId } ) {
 		return {
-			annotations: select( 'core/annotations' ).__experimentalGetAnnotationsForRichText( blockClientId, richTextIdentifier ),
+			annotations: select( STORE_KEY ).__experimentalGetAnnotationsForRichText( blockClientId, richTextIdentifier ),
 		};
 	},
-	__experimentalCreatePrepareEditableTree( props ) {
+	__experimentalCreatePrepareEditableTree( { annotations } ) {
 		return ( formats, text ) => {
-			if ( props.annotations.length === 0 ) {
+			if ( annotations.length === 0 ) {
 				return formats;
 			}
 
 			let record = { formats, text };
-			record = applyAnnotations( record, props.annotations );
+			record = applyAnnotations( record, annotations );
 			return record.formats;
 		};
 	},
+	__experimentalGetPropsForEditableTreeChangeHandler( dispatch ) {
+		return {
+			removeAnnotation: dispatch( STORE_KEY ).__experimentalRemoveAnnotation,
+			updateAnnotationRange: dispatch( STORE_KEY ).__experimentalUpdateAnnotationRange,
+		};
+	},
+	__experimentalCreateOnChangeEditableValue( props ) {
+		return ( formats ) => {
+			const positions = retrieveAnnotationPositions( formats );
+			const { removeAnnotation, updateAnnotationRange, annotations } = props;
+
+			updateAnnotationsWithPositions( annotations, positions, { removeAnnotation, updateAnnotationRange } );
```

Comment:
Why can `retrieveAnnotationPositions` and `updateAnnotationsWithPositions` not be done in one go, something like `updateAnnotationsFromFormats`?

Reply:
I felt like these two functions were good to make the code more understandable, they can be rolled into one if that's better.